Question: When to call `client.abort()` ?

I see there has some doc comments says need to call`client.abort()`:

https://github.com/obabec/rust-mqtt/blob/main/src/client/err.rs#L23

So I do this:

```rust
loop {
    match client.poll().with_timeout(keep_alive).await {
        Err(TimeoutError) => client.ping().await.or(Err(MqttUpError::Mqtt))?,
        Ok(Err(e)) => {
            info!("mqtt poll error: {}", e);
            if !e.is_recoverable() {
                client.abort().await; // panic at this line
            }
            return Err(MqttUpError::Mqtt);
        }
        Ok(Ok(Event::Publish(p))) => {
            let topic = p.topic.as_ref().as_str();
            let body = p.message.as_bytes();
            let body = str::from_utf8(body).unwrap();
            info!("received from broker, topic = {}, body = {}", topic, body);
        }
        Ok(Ok(_)) => {}
    }
}
```

Then:

```
[ERROR] panicked at 'Network must not be in Ok() state to disconnect due to an error.'
```

I read the source code in /rust-mqtt-0.5.1/src/client/raw/mod.rs :


```rust
// inside: pub async fn abort(&mut self) -> Result<(), RawError<B::ProvisionError>>
debug_assert!(
    !self.n.is_ok(),
    "Network must not be in Ok() state to disconnect due to an error."
);
```

but do not know how to check it should be call or not. Or just do not call it at all?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add request problem information support
 - Add reason string to incoming PUBACK, PUBREC, PUBREL, PUBCOMP, SUBACK, UNSUBACK
 - Detect protocol error when server sends user properties (only when `MAX_USER_PROPERTIES` > 0) or a reason string in the PUBACK, PUBREC, PUBREL, PUBCOMP, SUBACK and UNSUBACK packets
+- Fix client not dropping network connection on receival of a DISCONNECT packet leading to a panic on subsequent call to `Client::abort` in debug builds
 
 ## 0.5.1 - 2026-04-10
 

--- a/src/client/mod.rs
+++ b/src/client/mod.rs
@@ -1598,6 +1598,10 @@ impl<
                     .recv_body::<DisconnectPacket<MAX_USER_PROPERTIES>>(&header)
                     .await?;
 
+                // The server initiated the disconnect. We must close the transport on our side
+                // as well so that subsequent error handling (e.g. `abort`) sees a non-Ok network state.
+                self.raw.close_with(None);
+
                 return Err(MqttError::Disconnect {
                     reason: disconnect.reason_code,
                     reason_string: disconnect.reason_string.map(Property::into_inner),

--- a/tests/integration/connect_options.rs
+++ b/tests/integration/connect_options.rs
@@ -638,11 +638,15 @@ async fn keep_alive_not_kept_alive_incoming_qos0() {
 
         assert_ok!(
             timeout(Duration::from_secs(4), async {
-                loop {
-                    if let Err(MqttError::Network(_)) = rx.poll().await {
-                        break;
-                    }
-                }
+                while !matches!(
+                    rx.poll().await,
+                    Err(MqttError::Disconnect {
+                        reason: ReasonCode::KeepAliveTimeout,
+                        reason_string: _,
+                        user_properties: _,
+                        server_reference: _,
+                    }) | Err(MqttError::Network(_))
+                ) {}
             })
             .await,
             "expected to be disconnected"

--- a/tests/integration/connect_options.rs
+++ b/tests/integration/connect_options.rs
@@ -595,6 +595,9 @@ async fn keep_alive_not_kept_alive_idle_network() {
             server_reference: _,
         } | MqttError::Network(_)
     ));
+
+    // A non-recoverable poll error should always allow calling abort without panicking.
+    c.abort().await;
 }
 
 #[tokio::test]

--- a/tests/integration/connect_options.rs
+++ b/tests/integration/connect_options.rs
@@ -651,6 +651,8 @@ async fn keep_alive_not_kept_alive_incoming_qos0() {
             .await,
             "expected to be disconnected"
         );
+        // A non-recoverable poll error should always allow calling abort without panicking.
+        rx.abort().await;
     };
 
     join!(receiver, publisher);
@@ -702,6 +704,8 @@ async fn keep_alive_not_kept_alive_will_timing() {
             server_reference: _,
         } | MqttError::Network(_)
     ));
+    // A non-recoverable poll error should always allow calling abort without panicking.
+    tx.abort().await;
 
     disconnect(&mut rx, DEFAULT_DC_OPTIONS).await;
 }

